### PR TITLE
[OMNIML-2852] [2/n] Add Core Sparse Attention Infrastructure

### PR DESCRIPTION
**Pull request by kaix-nv**
## What does this PR do?

**Type of change:** ? 
New feature

**Overview:** ?
This PR provides a sparse attention support in ModelOpt for applying attention sparsity through skip softmax method, enabling inference speedups for LLMs.

Key Features:
- Skip softmax support
- Sparse attention config
- Extensible method registry for future sparse attention algorithms
- HuggingFace Transformers integration
- Phase-aware thresholds (separate prefill/decode)

[Design doc](https://docs.google.com/document/d/1OgmTAKkoD4ZSWYXel-FeaQqmI5PtyNhQ4dEuhGiZAQQ/edit?tab=t.0#heading=h.dyp44woziy9x)

## Usage

```python
import torch
import modelopt.torch.sparsity.attention_sparsity as mts
from transformers import AutoModelForCausalLM

# Load model (must use eager attention for softmax patching)
model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-hf",
    attn_implementation="eager",  # Required!
    torch_dtype=torch.bfloat16,
)

# Use pre-defined configuration
from modelopt.torch.sparsity.attention_sparsity import SKIP_SOFTMAX_DEFAULT
model = mts.sparsify(model, SKIP_SOFTMAX_DEFAULT)
```

### Unit Test

```bash
pytest tests/unit/torch/sparsity/attention_sparsity -v
pytest tests/gpu/torch/sparsity/attention_sparsity -v
pytest tests/examples/llm_sparsity/attention_sparsity -v
```
ALL PASSED.

### Accuracy
Benchmark: MMLU
Model: Qwen/Qwen3-4B
Cmd: python mmlu.py --model_name causal --model_path Qwen/Qwen3-4B --sparse_cfg SKIP_SOFTMAX_DEFAULT

|                      | MMLU  |
|----------------------|-------|
| BF16                 | 69.96 |
| SKIP_SOFTMAX_DEFAULT | 69.86 |

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

**Comment by cjluo-nv** [comment]
Hi @kaix-nv could you further split this code change? This PR has 3000+ lines of code change and many file moves

File: modelopt/torch/sparsity/attention_sparsity/sparse_attention.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            # Use sparse input if modified, otherwise use original
+            if sparse_input is not None:
+                return original_softmax(sparse_input, dim, *args, **kwargs)
+            return original_softmax(input, dim, *args, **kwargs)
+
+        return sparse_softmax
+
+
+# Create registry for sparse attention modules
+SparseAttentionRegistry = _DMRegistryCls("SparseAttention", SparseAttentionModule)
```
Comment (kevalmorabia97) [inline modelopt/torch/sparsity/attention_sparsity/sparse_attention.py:189]: Can we use a single registry for all Sparsity algorithms and modes and then use top-level `mts.sparsify(model, mode=...)` so all algorithms (e.g. weight or attention sparsify) are invoked by single shared API instead of separate API per algorithm?

File: tests/examples/llm_sparsity/attention_sparsity/test_attention_sparsity.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    )
+    kwargs.setdefault("seq_len", 128)
+    kwargs.setdefault("num_samples", 1)
+    kwargs.setdefault("max_new_tokens", 16)
+
+    cmd_parts = extend_cmd_parts(["python", "hf_sa.py"], **kwargs)
+    run_example_command(cmd_parts, "llm_sparsity/attention_sparsity")
+
+
+@minimum_gpu(1)
```
Comment (kevalmorabia97) [inline tests/examples/llm_sparsity/attention_sparsity/test_attention_sparsity.py]: No need for 1-gpu marker. All tests are run on 1 or more gpus only

File: tests/gpu/torch/sparsity/attention_sparsity/test_attention_sparsity_gpu.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    SimpleTransformerEncoderLayer,
+    get_test_configs,
+    save_restore_test,
+    sparsify_model_and_forward,
+)
+
+import modelopt.torch.sparsity.attention_sparsity as sparse_attn
+
+# Skip all tests if GPU is not available
+pytestmark = pytest.mark.skipif(not torch.cuda.is_available(), reason="GPU not available")
```
Comment (kevalmorabia97) [inline tests/gpu/torch/sparsity/attention_sparsity/test_attention_sparsity_gpu.py]: tests inside `tests/gpu` dont need a gpu check. Its assumed it is run only on gpu enabled machines. 

Same applies to all test files

File: tests/gpu/torch/sparsity/attention_sparsity/test_integration_gpu.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+@pytest.fixture(scope="module")
+def tiny_llama_dir(tmp_path_factory):
+    """Create minimal Llama model locally."""
+    return create_tiny_llama_dir(
+        tmp_path_factory.mktemp("tiny_llama"),
+        with_tokenizer=True,
+        num_hidden_layers=2,  # Minimal layers for fast testing
+        hidden_size=512,
+        intermediate_size=1024,
```
Comment (kevalmorabia97) [inline tests/gpu/torch/sparsity/attention_sparsity/test_integration_gpu.py]: why do we need such large hidden and intermediate size? Can we use 32/64 instead?

File: tests/unit/torch/sparsity/attention_sparsity/test_sparse_attention_config.py
Comment (kevalmorabia97) [inline tests/unit/torch/sparsity/attention_sparsity/test_sparse_attention_config.py:1]: these tests seem unnecessary

File: modelopt/torch/sparsity/attention_sparsity/plugins/huggingface.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            if module_type not in attention_types:
+                SparseAttentionRegistry.register({module_type: type_name})(_GenericSparseAttention)
+                attention_types.add(module_type)
+                registered_count += 1
+                logger.info(f"Registered {type_name} for sparse attention optimization")
+
+    if registered_count > 0:
+        logger.info(
+            f"Dynamically registered {registered_count} attention module types for sparsity"
+        )
```
Comment (jy-yuan) [inline modelopt/torch/sparsity/attention_sparsity/plugins/huggingface.py]: Consider only using `logger.info()` or `print()` for consistency? Some other places in this pr are using `print()`.

File: modelopt/torch/sparsity/attention_sparsity/methods/flash_skip_softmax.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+
+            # Step 8: Calculate sparsity statistics
+            # Count kept blocks (averaged across batch and heads)
+            kept_blocks = block_mask.sum().item() / (batch_size * num_heads)
+
+            # Total valid blocks (lower triangle only for causal attention)
+            # Note: Causal mask pre-applied by attention module, so block_mask naturally
+            # has zeros in upper triangle. We only count lower triangle for denominator.
+            total_blocks = (
+                num_block_rows * (num_block_rows + 1) // 2  # Causal: N(N+1)/2
```
Comment (jy-yuan) [inline modelopt/torch/sparsity/attention_sparsity/methods/flash_skip_softmax.py:194]: Does that means rows==columns? Which means we only have causal in self-attention, not cross-attention?

File: examples/llm_sparsity/attention_sparsity/hf_sa.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        type=str,
+        default="skip_softmax",
+        choices=list(SPARSE_ATTN_CFG_CHOICES.keys()),
+        help="Sparse attention configuration to apply.",
+    )
+    parser.add_argument(
+        "--backend",
+        type=str,
+        default="pytorch",
+        choices=["pytorch", "triton"],
```
Comment (jy-yuan) [inline examples/llm_sparsity/attention_sparsity/hf_sa.py]: Is `"triton"` a TODO?

File: modelopt/torch/sparsity/attention_sparsity/conversion.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+    print(f"Disabled: {disabled_count}")
+
+    if method_counts:
+        print("\nMethods:")
+        for method, count in sorted(method_counts.items()):
+            print(f"{method}: {count}")
+
+    for name, module in sparse_modules:
+        method = getattr(module, "_method", "unknown")
+        threshold = getattr(module, "_threshold", "N/A")
```
Comment (jy-yuan) [inline modelopt/torch/sparsity/attention_sparsity/conversion.py]: Do `SparseAttentionModule` have `_method` or `_threshold`?

File: modelopt/torch/sparsity/attention_sparsity/conversion.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+                # Restore method and config
+                if "method" in module_state:
+                    module._method = module_state["method"]
+                if "method_config" in module_state:
+                    # Restore config attributes
+                    for key, val in module_state["method_config"].items():
+                        setattr(module, f"_{key}", val)
+
+                # Re-setup with restored config
+                module._setup()
```
Comment (jy-yuan) [inline modelopt/torch/sparsity/attention_sparsity/conversion.py:210]: Need add test for this?

File: examples/llm_sparsity/attention_sparsity/hf_sa.py (hunk 7 above)
Comment (kaix-nv) [reply examples/llm_sparsity/attention_sparsity/hf_sa.py]: yes

File: modelopt/torch/sparsity/attention_sparsity/conversion.py (hunk 8 above)
Comment (kaix-nv) [reply modelopt/torch/sparsity/attention_sparsity/conversion.py]: `print_sparse_attention_summary` isn’t used in this PR, I’ve removed it. It will be introduced in the next PR.

File: modelopt/torch/sparsity/attention_sparsity/conversion.py (hunk 9 above)
Comment (kaix-nv) [reply modelopt/torch/sparsity/attention_sparsity/conversion.py:210]: `test_restore_sparse_attention_model` covers the test for this func.

File: modelopt/torch/sparsity/attention_sparsity/sparse_attention.py (hunk 1 above)
Comment (kaix-nv) [reply modelopt/torch/sparsity/attention_sparsity/sparse_attention.py:189]: This is a good advice. I'll submit a follow-up PR later.

File: modelopt/torch/sparsity/attention_sparsity/methods/flash_skip_softmax.py (hunk 6 above)
Comment (kaix-nv) [reply modelopt/torch/sparsity/attention_sparsity/methods/flash_skip_softmax.py:194]: Yes, it's for causal attention in prefill.

**Comment by kaix-nv** [comment]
@kevalmorabia97 I've addressed the review suggestions. Could you please review and approve the PR so I can move forward with the subsequent PRs? Thanks.

File: examples/llm_sparsity/weight_sparsity/finetune.py
Comment (kevalmorabia97) [inline examples/llm_sparsity/weight_sparsity/finetune.py:1]: This license is already present below stanford_alpaca's license. Please fix the path in https://github.com/NVIDIA/Model-Optimizer/blob/main/.pre-commit-config.yaml#L95 so pre-commit doesnt add the license again

**Review by kevalmorabia97 — COMMENTED** [review]
Only reviewed high-level structure. Would suggest someone else take a look at the core sparsity logic as I'm not familiar with it

**Comment by kaix-nv** [comment]
> Only reviewed high-level structure. Would suggest someone else take a look at the core sparsity logic as I'm not familiar with it
I've asked @jy-yuan to review since he's very familiar with the core logic.
@jy-yuan Please approve if you think the PR is in good shape. Thanks.

**Comment by realAsma** [comment]
Looks great! 

Should we have a simpler high-level usage which aligns with `mtq`?
```
# Use pre-defined configuration
model = mts.sparsify(model, mts.SPARSE_ATTEN_SKIP_SOFTMAX_CFG)
```

**Comment by jy-yuan** [comment]
> > Only reviewed high-level structure. Would suggest someone else take a look at the core sparsity logic as I'm not familiar with it
> > I've asked @jy-yuan to review since he's very familiar with the core logic.
> > @jy-yuan Please approve if you think the PR is in good shape. Thanks.

I reviewed and approve, thanks! @kevalmorabia97

**Comment by kaix-nv** [comment]
> Looks great!
> 
> Should we have a simpler high-level usage which aligns with `mtq`?
> 
> ```
> # Use pre-defined configuration
> model = mts.sparsify(model, mts.SPARSE_ATTEN_SKIP_SOFTMAX_CFG)
> ```

Good point. Initially, I wanted to separate attention sparsity from weight sparsity, so I choose different APIs for each feature. But using a single API is indeed more consistent, I’ll submit a follow-up PR to unify them.

File: examples/llm_sparsity/weight_sparsity/finetune.py
Comment (kevalmorabia97) [inline examples/llm_sparsity/weight_sparsity/finetune.py:1]: Please revert license header diff in this file
